server/websession.py

```diff
@@ -0,0 +1,97 @@
+# code for this started by copying the bits of flask-session
+# https://github.com/mcrowson/flask-sessionstore-fork/blob/master/flask_sessionstore/sessions.py
+# that are relevant for database-backed storage, then simplifying for just our use case.
+
+import secrets
+from datetime import datetime, timezone
+
+from flask.sessions import SessionInterface, SessionMixin
+from werkzeug.datastructures import CallbackDict
+from .models import WebSession
+from .database import db_session
+from . import config
+
+
+class ArloSession(CallbackDict, SessionMixin):  # pylint: disable=too-many-ancestors
+    def __init__(self, sid=None, initial=None):
+        def on_update(self):
+            self.modified = True
+
+        CallbackDict.__init__(self, initial, on_update)
+        self.sid = sid
+        self.permanent = True
+        self.modified = False
+
+
+class ArloSessionInterface(SessionInterface):
+    def _generate_sid(self):
+        return secrets.token_urlsafe(50)
+
+    def open_session(self, app, request):
+        sid = request.cookies.get(app.session_cookie_name)
+        if not sid:
+            return ArloSession(sid=self._generate_sid())
+
+        saved_session = WebSession.query.filter_by(id=sid).first()
+
+        if saved_session:
+            return ArloSession(sid=sid, initial=saved_session.data)
+        else:
+            return ArloSession(sid=self._generate_sid())
+
+    def save_session(self, app, session, response):
+        if not session.modified:
+            return
+
+        # we're about to update the database, and we want to cancel any pending and uncommitted updates.
+        # because we use scoped_session, which maintains a thread-local sqlalchemy session,
+        # calling remove here tidies everything up, including potentially rolling back an uncommitted txn,
+        # and on next db_session call, we'll be using a new thread-local session.
+        db_session.remove()
```

Comment:
What's an example of a pending/uncommitted update that would be happening when `save_session` is called? Just want to get an intuition for why this is useful

Reply:
Ah ok so it's specific to when flask actually calls these methods. That's quite subtle. Could you add a bit of context around that in the comment here explaining it?

Also, just wondering if we're sure we want to run `db_session.commit()` in `save_session`. Would it break anything to make the updating of the web session part of the same transaction as the rest of the endpoint handler?

Reply:
Is there a different place we would call `db_session.commit()` ? My understanding is if we don't call it here, then the handler from `app.py` will roll back, thus preventing an update. I believe this approach follows the pattern we currently use, where the specific URL handler is responsible for committing its own data, and so is this session handler.

But, open to other thoughts!

Reply:
We could change the `app.teardown_context` handler to be a bit smarter and only rollback on error, otherwise commit (see https://flask.palletsprojects.com/en/1.1.x/api/#flask.Flask.teardown_appcontext). That would ensure that every request ends in a transaction commit (we could even use that to avoid having to manually commit in all of our endpoint handlers!)

I think we should think through what behavior we want - i.e. ideally, when should the session update be committed to the database? How coupled should it be to the success of the request?

Reply:
My immediate reaction to that idea is negative, but I would be happy to discuss it.

Either way, though, I think there's a good case to be made for the session database commit to be independent of the business logic commit. You might want to save an error message in the sessions state, for example. In fact I tried to figure out how to make the session store's database connection a *different* one, under separate transaction control, but `scoped_session` appears to be too smart and seems to prevent me from doing this, or at least I couldn't figure it out. But if I had my way, I'd make the session's database connection independent of the URL handler's.

Reply:
I buy that it should be independent of the business logic, since the request happened whether or not it succeeded.